### PR TITLE
Derandomized smoothing

### LINKED ISSUE #1613: Implementation of (De)Randomized Smoothing
(De)Randomized Smoothing,  https://arxiv.org/abs/2002.10733, is a simple yet highly effective method for certifying against patch attacks in neural networks. It has also recently been extended and shown to be very effective on Transformers (https://arxiv.org/abs/2110.07719) due to their ability to handle large masked images. Adding this feature to ART will be a valuable and important step in expanding the suite of certification methods offered.

### PR DESCRIPTION
**Pull request by GiulioZizzo**
# Description

We implement derandomized smoothing, a method to certify neural networks against patch based attacks.

Fixes  #1613

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- We introduce tests in test_derandomized_smoothing.py to verify the certified performance on MNIST. For CIFAR we just check that the code runs.

**Test Configuration**:
- RHEL8
- Python version: 3.9
- ART version or commit number: 1.10.1
- TensorFlow: 2.8.0/ / PyTorch: 1.11.0 
- 
# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+    """
+
+    def __init__(
+        self,
+        ablation_type: str,
+        ablation_size: int,
+        threshold: float,
+        logits: bool,
+        channels_first: bool,
+        *args,
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:49]:
```suggestion
```

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+
+            x[:, :, :row_pos, :] = 0.0
+            x[:, :, row_pos + k :, :] = 0.0
+        else:
+            raise ValueError(f"Ablation failed on row: {row_pos} and column: {column_pos} with size {k}")
+
+        return x
+
+
+ABLATOR_TYPE = Union[BlockAblator, ColumnAblator]  # pylint: disable=C0103
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:445]: Is it possible to move this type to `art.utils` to the other type definitions?

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        Perform prediction for a batch of inputs.
+
+        :param x: Input samples.
+        :param batch_size: Size of batches.
+        :param training_mode: `True` for model set to training mode and `'False` for model set to evaluation mode.
+        :return: Array of predictions of shape `(nb_inputs, nb_classes)`.
+        """
+        raise NotImplementedError
+
+    def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> np.ndarray:  # pylint: disable=W0613
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:91]: Which argument is unused causing W0613?

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+"""
+This module implements (De)Randomized Smoothing for Certifiable Defense against Patch Attacks
+
+| Paper link: https://arxiv.org/abs/2002.10733
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from abc import ABC, abstractmethod
+
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:27]:
```suggestion
from abc import ABC, abstractmethod
```

File: art/estimators/certification/derandomized_smoothing/__init__.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+"""
+Randomized smoothing estimators.
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/__init__.py:2]: Is it DeRandomized or Randomized?
```suggestion
DeRandomized smoothing estimators.
```

File: art/estimators/certification/derandomized_smoothing/pytorch.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/pytorch.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: art/estimators/certification/derandomized_smoothing/pytorch.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import List, Optional, Tuple, Union, Any, TYPE_CHECKING
+import random
+
+import numpy as np
+
+from tqdm import tqdm
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/pytorch.py:32]:
```suggestion
import numpy as np
from tqdm import tqdm
```

File: art/estimators/certification/derandomized_smoothing/tensorflow.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Callable, List, Optional, Tuple, Union, TYPE_CHECKING
+
+import numpy as np
+import tensorflow as tf
+
+from tqdm import tqdm
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/tensorflow.py:31]:
```suggestion
import numpy as np
import tensorflow as tf
from tqdm import tqdm
```

File: tests/estimators/certification/test_derandomized_smoothing.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    TensorFlowV2DeRandomizedSmoothing,
+)
+from tests.utils import ARTTestException
+
+
+@pytest.fixture()
+def fix_get_mnist_data():
+    """
+    Get the first 128 samples of the mnist test set with channels first format
+    :return: First 128 sample/label pairs of the MNIST test dataset.
```
Comment (beat-buesser) [inline tests/estimators/certification/test_derandomized_smoothing.py:35]:
```suggestion
    Get the first 128 samples of the mnist test set with channels first format

    :return: First 128 sample/label pairs of the MNIST test dataset.
```

File: tests/estimators/certification/test_derandomized_smoothing.py [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+
+    x_test, y_test = x_test[:nb_test], y_test[:nb_test]
+    return x_test, y_test
+
+
+@pytest.fixture()
+def fix_get_cifar10_data():
+    """
+    Get the first 128 samples of the cifar10 test set
+    :return: First 128 sample/label pairs of the cifar10 test dataset.
```
Comment (beat-buesser) [inline tests/estimators/certification/test_derandomized_smoothing.py:52]:
```suggestion
    Get the first 128 samples of the cifar10 test set

    :return: First 128 sample/label pairs of the cifar10 test dataset.
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @GiulioZizzo Thank you very much for your pull request contributing de-randomised smoothing! The code looks good and I have a left a few comments and questions on formatting. What do you think?

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py (hunk 3 above)
Comment (GiulioZizzo) [reply art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:94]: Hi Beat, the **kwargs argument is unused to keep the same function signature as other predict methods.

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py (hunk 2 above)
Comment (GiulioZizzo) [reply art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py]: I agree, moved it!

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py (hunk 4 above)
Comment (GiulioZizzo) [reply art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:27]: Hi Beat, it looks like there aren't any suggested changes here.

Comment (beat-buesser) [reply art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:27]: Sorry, I wanted to suggest to remove line 27 between the imports of `abc` and `typing` to have them in one block.

File: art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py (hunk 3 above)
Comment (beat-buesser) [reply art/estimators/certification/derandomized_smoothing/derandomized_smoothing.py:95]: Ok, I see. How about if we pass `kwargs` to the calls of _predict_classifier`?

File: art/estimators/certification/derandomized_smoothing/tensorflow.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline art/estimators/certification/derandomized_smoothing/tensorflow.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

**Review by beat-buesser — APPROVED** [review]
@GiulioZizzo Thank you!
